@vthapar the Helm job is now failing on some ServiceDiscovery use-case (the previous RBAC issues are now resolved). Can you please take a look.
```
STEP: Creating serviceExport nginx-demo.e2e-tests-discovery-698nz on "cluster2"
STEP: Retrieving ServiceExport nginx-demo.e2e-tests-discovery-698nz on "cluster2"
panic: test timed out after 15m0s

goroutine 818 [running]:
testing.(*M).startAlarm.func1()
```

--- a/test/e2e/discovery/headless_services.go
+++ b/test/e2e/discovery/headless_services.go
@@ -39,41 +39,25 @@ var _ = Describe("[discovery] Test Headless Service Discovery Across Clusters", 
 
 	When("a pod tries to resolve a headless service in a remote cluster", func() {
 		It("should resolve the backing pod IPs from the remote cluster", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunHeadlessDiscoveryTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunHeadlessDiscoveryTest(f)
 		})
 	})
 
 	When("a pod tries to resolve a headless service which is exported locally and in a remote cluster", func() {
 		It("should resolve the backing pod IPs from both clusters", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunHeadlessDiscoveryLocalAndRemoteTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunHeadlessDiscoveryLocalAndRemoteTest(f)
 		})
 	})
 
 	When("the number of active pods backing a service changes", func() {
 		It("should only resolve the IPs from the active pods", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunHeadlessPodsAvailabilityTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunHeadlessPodsAvailabilityTest(f)
 		})
 	})
 
 	When("a pod tries to resolve a headless service in a specific remote cluster by its cluster name", func() {
 		It("should resolve the backing pod IPs from the specified remote cluster", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunHeadlessDiscoveryClusterNameTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunHeadlessDiscoveryClusterNameTest(f)
 		})
 	})
 })
@@ -96,8 +80,7 @@ func RunHeadlessDiscoveryTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipList, hostNameList := f.GetEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace)
-
+	ipList, hostNameList := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
 		"", true)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
@@ -143,8 +126,8 @@ func RunHeadlessDiscoveryLocalAndRemoteTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipListB, hostNameListB := f.GetEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace)
-	ipListA, hostNameListA := f.GetEndpointIPs(framework.ClusterA, nginxHeadlessClusterA.Name, nginxHeadlessClusterA.Namespace)
+	ipListB, _ := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
+	ipListA, _ := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
 	ipList := make([]string, len(ipListB)+len(ipListA))
 	ipList = append(ipList, ipListB...)
 	ipList = append(ipList, ipListA...)
@@ -157,12 +140,13 @@ func RunHeadlessDiscoveryLocalAndRemoteTest(f *lhframework.Framework) {
 
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListB, checkedDomains,
 		"", false)
-	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListA, checkedDomains,
-		"", true)
+	/* TODO: SRV query without cluster is broken. Uncomment once fixed
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
 		clusterBName, true, false, true)
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-		clusterBName, true, false, true)
+		clusterBName, true, false, true)*/
+	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListA, checkedDomains,
+		"", true)
 }
 
 func RunHeadlessPodsAvailabilityTest(f *lhframework.Framework) {
@@ -185,17 +169,17 @@ func RunHeadlessPodsAvailabilityTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipList, _ := f.AwaitEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace, 3)
+	ipList, _ := f.AwaitPodIPs(framework.ClusterB, nginxHeadlessClusterB, 3)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
 		"", true)
 
 	f.SetNginxReplicaSet(framework.ClusterB, 0)
-	ipList, _ = f.AwaitEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace, 0)
+	ipList, _ = f.AwaitPodIPs(framework.ClusterB, nginxHeadlessClusterB, 0)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
 		"", false)
 
 	f.SetNginxReplicaSet(framework.ClusterB, 2)
-	ipList, _ = f.AwaitEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace, 2)
+	ipList, _ = f.AwaitPodIPs(framework.ClusterB, nginxHeadlessClusterB, 2)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
 		"", true)
 }
@@ -228,19 +212,22 @@ func RunHeadlessDiscoveryClusterNameTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipListClusterA, hostNameListA := f.GetEndpointIPs(framework.ClusterA, nginxHeadlessClusterA.Name, nginxHeadlessClusterA.Namespace)
-	ipListClusterB, hostNameListB := f.GetEndpointIPs(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace)
+	ipListClusterA, _ := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
+	ipListClusterB, hostNameListB := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
 
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterA, netshootPodList, ipListClusterA, checkedDomains,
 		clusterAName, true)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListClusterB, checkedDomains,
 		clusterBName, true)
-	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-		clusterBName, true, true, true)
+	// TODO: Uncomment following checks once fixed
+	// verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+	//	clusterBName, true, true, true)
+
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
 		clusterBName, true, true, true)
-	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-		clusterBName, false, true, true)
+	// TODO: Uncomment following checks once fixed
+	// verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+	//	clusterBName, false, true, true)
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
 		clusterBName, false, true, true)
 }
@@ -340,11 +327,11 @@ func verifyHeadlessSRVRecordsWithDig(f *framework.Framework, cluster framework.C
 							strings.Contains(result.(string), hostDNS)
 					}
 					if doesContain && !shouldContain {
-						return false, fmt.Sprintf("expected execution result %q not to contain %q", result, hostName), nil
+						return false, fmt.Sprintf("expected execution result %q not to contain %q and %d", result, hostDNS, int(port.Port)), nil
 					}
 
 					if !doesContain && shouldContain {
-						return false, fmt.Sprintf("expected execution result %q to contain %q", result, hostName), nil
+						return false, fmt.Sprintf("expected execution result %q to contain %q and %d", result, hostDNS, int(port.Port)), nil
 					}
 				}
 
@@ -372,5 +359,5 @@ func createSRVQuery(f *framework.Framework, port corev1.ServicePort, service *co
 
 	cmd = append(cmd, portDNS)
 
-	return cmd, domainName
+	return cmd, fmt.Sprintf("%s.%s", clusterName, domainName)
 }

--- a/test/e2e/discovery/headless_services.go
+++ b/test/e2e/discovery/headless_services.go
@@ -126,8 +126,8 @@ func RunHeadlessDiscoveryLocalAndRemoteTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipListB, _ := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
-	ipListA, _ := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
+	ipListB, hostNameListB := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
+	ipListA, hostNameListA := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
 	ipList := make([]string, len(ipListB)+len(ipListA))
 	ipList = append(ipList, ipListB...)
 	ipList = append(ipList, ipListA...)
@@ -135,18 +135,22 @@ func RunHeadlessDiscoveryLocalAndRemoteTest(f *lhframework.Framework) {
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipList, checkedDomains,
 		"", true)
 
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
+		clusterBName, true, false, true)
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+		clusterAName, true, false, true)
 	f.DeleteServiceExport(framework.ClusterB, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace)
 	f.AwaitServiceImportCount(framework.ClusterA, nginxHeadlessClusterB.Name, nginxHeadlessClusterB.Namespace, 1)
 
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListB, checkedDomains,
 		"", false)
-	/* TODO: SRV query without cluster is broken. Uncomment once fixed
-	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
-		clusterBName, true, false, true)
-	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-		clusterBName, true, false, true)*/
+
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListA, checkedDomains,
 		"", true)
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
+		clusterBName, true, false, false)
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+		clusterAName, true, false, true)
 }
 
 func RunHeadlessPodsAvailabilityTest(f *lhframework.Framework) {
@@ -212,22 +216,21 @@ func RunHeadlessDiscoveryClusterNameTest(f *lhframework.Framework) {
 
 	netshootPodList := f.NewNetShootDeployment(framework.ClusterA)
 
-	ipListClusterA, _ := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
+	ipListClusterA, hostNameListA := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
 	ipListClusterB, hostNameListB := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
 
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterA, netshootPodList, ipListClusterA, checkedDomains,
 		clusterAName, true)
 	verifyHeadlessIpsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, ipListClusterB, checkedDomains,
 		clusterBName, true)
-	// TODO: Uncomment following checks once fixed
-	// verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-	//	clusterBName, true, true, true)
 
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+		clusterAName, true, true, true)
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
 		clusterBName, true, true, true)
-	// TODO: Uncomment following checks once fixed
-	// verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
-	//	clusterBName, false, true, true)
+
+	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListA, checkedDomains,
+		clusterAName, false, true, true)
 	verifyHeadlessSRVRecordsWithDig(f.Framework, framework.ClusterA, nginxHeadlessClusterB, netshootPodList, hostNameListB, checkedDomains,
 		clusterBName, false, true, true)
 }
@@ -318,14 +321,8 @@ func verifyHeadlessSRVRecordsWithDig(f *framework.Framework, cluster framework.C
 				}
 				for _, hostName := range hostNameList {
 					hostDNS := hostName + "." + domainName
-					var doesContain bool
-					if shouldContain {
-						doesContain = strings.Contains(result.(string), strconv.Itoa(int(port.Port))) &&
-							strings.Contains(result.(string), hostDNS)
-					} else {
-						doesContain = strings.Contains(result.(string), strconv.Itoa(int(port.Port))) ||
-							strings.Contains(result.(string), hostDNS)
-					}
+					doesContain := strings.Contains(result.(string), strconv.Itoa(int(port.Port))) &&
+						strings.Contains(result.(string), hostDNS)
 					if doesContain && !shouldContain {
 						return false, fmt.Sprintf("expected execution result %q not to contain %q and %d", result, hostDNS, int(port.Port)), nil
 					}

--- a/test/e2e/discovery/service_discovery.go
+++ b/test/e2e/discovery/service_discovery.go
@@ -92,12 +92,6 @@ var _ = Describe("[discovery] Test Service Discovery Across Clusters", func() {
 				return
 			}
 
-			// TODO: Enable this once globalnetv2 fixes health check
-			if framework.TestContext.GlobalnetEnabled {
-				Skip("Not supported with Globalnet v2 yet and hence skipping the test")
-				return
-			}
-
 			randomIP := "192.168.1.5"
 			endpointName, healthCheckIP = f.GetHealthCheckIPInfo(framework.ClusterC)
 			f.SetHealthCheckIP(framework.ClusterC, randomIP, endpointName)

--- a/test/e2e/discovery/statefulsets.go
+++ b/test/e2e/discovery/statefulsets.go
@@ -35,31 +35,19 @@ var _ = Describe("[discovery] Test Stateful Sets Discovery Across Clusters", fun
 
 	When("a pod tries to resolve a podname from stateful set in a remote cluster", func() {
 		It("should resolve the pod IP from the remote cluster", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunSSDiscoveryTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunSSDiscoveryTest(f)
 		})
 	})
 
 	When("a pod tries to resolve a podname from stateful set in a local cluster", func() {
 		It("should resolve the pod IP from the local cluster", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunSSDiscoveryLocalTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunSSDiscoveryLocalTest(f)
 		})
 	})
 
 	When("the number of active pods backing a stateful set changes", func() {
 		It("should only resolve the IPs from the active pods", func() {
-			if !framework.TestContext.GlobalnetEnabled {
-				RunSSPodsAvailabilityTest(f)
-			} else {
-				framework.Skipf("Globalnet is enabled, skipping the test...")
-			}
+			RunSSPodsAvailabilityTest(f)
 		})
 	})
 })

--- a/test/e2e/framework/framework.go
+++ b/test/e2e/framework/framework.go
@@ -287,7 +287,12 @@ func (f *Framework) AwaitEndpointIPs(targetCluster framework.ClusterIndex, name,
 		for _, eps := range endpoint.Subsets {
 			for _, addr := range eps.Addresses {
 				ipList = append(ipList, addr.IP)
-				hostNameList = append(hostNameList, addr.Hostname)
+				switch {
+				case addr.Hostname != "":
+					hostNameList = append(hostNameList, addr.Hostname)
+				case addr.TargetRef != nil:
+					hostNameList = append(hostNameList, addr.TargetRef.Name)
+				}
 			}
 		}
 		if count != anyCount && len(ipList) != count {
@@ -299,8 +304,36 @@ func (f *Framework) AwaitEndpointIPs(targetCluster framework.ClusterIndex, name,
 	return ipList, hostNameList
 }
 
-func (f *Framework) GetEndpointIPs(targetCluster framework.ClusterIndex, name, namespace string) (ipList, hostNameList []string) {
-	return f.AwaitEndpointIPs(targetCluster, name, namespace, anyCount)
+func (f *Framework) AwaitPodIngressIPs(targetCluster framework.ClusterIndex, svc *v1.Service, count int) (ipList, hostNameList []string) {
+	podList := f.Framework.AwaitPodsByAppLabel(targetCluster, svc.Labels["app"], svc.Namespace, count)
+	hostNameList = make([]string, 0)
+	ipList = make([]string, 0)
+
+	for i := 0; i < len(podList.Items); i++ {
+		ingressIPName := fmt.Sprintf("pod-%s", podList.Items[i].Name)
+		ingressIP := f.Framework.AwaitGlobalIngressIP(targetCluster, ingressIPName, svc.Namespace)
+		ipList = append(ipList, ingressIP)
+		hostname := podList.Items[i].Spec.Hostname
+		if hostname == "" {
+			hostname = podList.Items[i].Name
+		}
+
+		hostNameList = append(hostNameList, hostname)
+	}
+
+	return ipList, hostNameList
+}
+
+func (f *Framework) AwaitPodIPs(targetCluster framework.ClusterIndex, svc *v1.Service, count int) (ipList, hostNameList []string) {
+	if framework.TestContext.GlobalnetEnabled {
+		return f.AwaitPodIngressIPs(targetCluster, svc, count)
+	}
+
+	return f.AwaitEndpointIPs(targetCluster, svc.Name, svc.Namespace, count)
+}
+
+func (f *Framework) GetPodIPs(targetCluster framework.ClusterIndex, service *v1.Service) (ipList, hostNameList []string) {
+	return f.AwaitPodIPs(targetCluster, service, anyCount)
 }
 
 func (f *Framework) SetNginxReplicaSet(cluster framework.ClusterIndex, count uint32) *appsv1.Deployment {

